core/src/main/java/com/scalar/db/api/Scan.java

```diff
@@ -52,6 +57,27 @@ public Scan(Scan scan) {
     limit = scan.limit;
   }
 
+  /**
+   * Build a {@code Scan} or {@code ScanAll} operation using a builder
+   *
+   * @return a {@code Scan} operation builder
+   */
+  public static Namespace newBuilder() {
+    return new Namespace();
+  }
+
+  /**
+   * Build a {@code Scan} operation from an existing {@code Scan} object using a builder. The
+   * builder will be parametrized by default with all the existing {@code Scan} attributes
+   *
+   * @param scan an existing {@code Scan} operation
+   * @return a {@code Scan} operation builder
+   */
+  public static BuildableScanOrScanAllFromExisting newBuilder(Scan scan) {
+    checkNotNull(scan);
+    return new BuildableScanOrScanAllFromExisting(scan);
+  }
```

Comment:
The builder from an existing Scan is a bit particular as we wanted to expose a single `Scan.newBuilder(scan)` method to build either Scan or ScanAll object, while making the builder exposes as much as possible only dedicated Scan or ScanAll methods.

**Building from an existing ScanAll**
Calling the `Scan.newBuilder(existingScanAll).all()` method switches the builder into a `ScanAll` building mode, exposing only ScanAll related methods. Only then can the `existingScanAll` object can be modified and built
```
Scan.newBuilder(existingScanAll).all().table().build(); --> nominal behavior
Scan.newBuilder(existingScanAll).all().forPartittionKey().build(); --> by design, it is not possible to write this
Scan.newBuilder(existingScan).all().table().build(); -->will raise an IllegalStateException as existingScan variable is not a ScanAll

```
Calling any other methods besides `Scan.newBuilder(existingScanAll).all()` will raise an IllegalStateException
```
Scan.newBuilder(existingScanAll).forPartitionKey(); --> raise IllegalStateException
Scan.newBuilder(existingScanAll).table(); --> raise IllegalStateException
```
**Building from an existing Scan**
Calling any of the methods besides `Scan.newBuilder(existingScan).all()` will switch the builder into a `Scan` building mode, exposing only Scan-related methods.
```
Scan.newBuilder(existingScan).table().build(); --> nominal behavior
Scan.newBuilder(existingScan).table().all() --> by design, it is not possible to write this;
Scan.newBuilder(existingScan).all() --> will raise  an IllegalStateException
```

Reply:
For `ScanAll`, if `all()` is called right after `newBuilder()` when building from `existingScanAll`, it's more consistent that `all()` is called right after `newBuilder()` when building from scratch too?

```
// Build ScanAll from existing ScanAll
Scan.newBuilder(existingScanAll).all().table().build();

// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().partitionKey().build();
```

What do you think?

Reply:
Ok, in that case putting the `partitionKey()` method for a regular Scan right after the `newBuilder()` is probably the most intuitive. `Scan.newBuilder()` will then expose `all()` and `partitionKey()`.
```
// Build ScanAll from existing ScanAll
Scan.newBuilder(existingScanAll).all().table().build();

// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().build();

// Build regular Scan from scratch
Scan.newBuilder().partitionKey().namespace().table().build();
```
What do you think ?

Reply:
I thought the following is more consistent and intuitive:
```
// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().build();

// Build regular Scan from scratch
Scan.newBuilder().namespace().table().partitionKey().ordering().start().build();
```

So you mean `all()` is like `allPartitionKeys()`?

Reply:
After discussing with @brfrn169, we decided to keep the current implementation when building a Scan from scratch. 
```
//ScanAll
Scan.newBuilder().namespace().table().all().table().build();
//Scan
Scan.newBuilder().namespace().table().partitionKey().limit().build()
```
And revise the Scan builder from an existing object with the following 
 - remove the `Scan.newBuilder(existingScanAll).all()` method
 - `Scan.newBuilder(existingScanorScanAll)` will expose all the methods for Scan
 - when building from an existing ScanAll, if unsupported methods (such as partitionKey(),start(),etc...) are used, it will throw an exception. 
 For example 
 ```
 Scan.newBuilder(existingScanAll).partitionKey(); ->will throw UnsupportedOperationException
 Scan.newBuilder(existingScanAll).start(); ->will throw UnsupportedOperationException
 ```

Reply:
Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/3d59cd43e6820b848e514c74bddd25538c33be30